Define skeleton for new WordPressData module

mokagio opened the pull request:
Internal ref pbArwn-7b1-p2

- Defines an Objective-C + Swift combo package "WordPressData" where we'll progressively house _all_ of the model definition, storage, and services
- Move `CoreDataStack(Swift)` in WordPressData

**Testing**

- See green CI as a proof that the compilation is successful
- Try out the prototype builds, I guess

**Next steps**

- Move the `xcdatamodeld` folder into the package. Also check out how [Tracks](https://github.com/Automattic/Automattic-Tracks-iOS/tree/437f586a62c07c6ceec9baf043237b2afe7ea1ac/Sources/Model/ObjC/Common/Core%20Data) does it
  - Init the Core Data stack using the model from the package
- Move all the models (I have a feeling they need to be moved in bulk, but I might be wrong of course) to the package
- Move types heavily related with Core Data, such as `TaggedManagedObjectID` into the package

Looking at the code, I noticed two things that we need to take a decision on:

- Some of the code uses CocoaLumberjack. _Ditch the logs, add it as a dependency, or implement an abstraction and injection system like we did for WordPressAuthenticator?_
- Some of the code uses `WPException`. _Make it into its own package to import it in WordPressData and in the other code using `WPException`, or expose it through WordPressData and make those types that need it import WordPressData._

<details>
<summary>My thoughts in the details. What do you think @crazytonyli  @kean ?</summary>

- Some of the code uses CocoaLumberjack. As far as I know we don't use those logs much. If that's correct, let's keep things simple and remove the dependency.
- Some of the code uses `WPException`. Packages are cheap to add thanks to the current setup. Let's keep things clean and make a micro-package for `WPException`. It also keeps things clearer because importing a model/data package only to use an exception seems confusing to me.

</details>

## Regression Notes
1. Potential unintended areas of impact — _Anything that has to do with data storage in the app..._
2. What I did to test those areas of impact (or what existing automated tests I relied on) — At this time, only a quick smoke tests of the prototype builds, plus relied on the existing unit tests
3. What automated tests I added (or what prevented me from doing so) — None, as I only moved code at this time.

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes. — N.A.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

Testing checklist:
- [ ] WordPress.com sites and self-hosted Jetpack sites.
- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] VoiceOver.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] iPhone and iPad. 
- [ ] Multi-tasking: Split view and Slide over. (iPad)

mokagio committed:
Move `CoreDataStack` to new `WordPressData` module

dangermattic commented:
<!--
  0 Errors
  2 Warnings: <code>Modules/Package.swift</c...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><code>Modules/Package.swift</code> was changed without updating its corresponding <code>Package.resolved</code>. Please resolve the Swift packages as appropriate to your project setup (e.g. in Xcode or by running <code>swift package resolve</code>).</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot or video is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

mokagio committed:
Add `@_exported import` for `WordPressDataObjC` in `WordPressData`

mokagio committed:
Add a dummy file to WordPressDataObjC so its .o can be generated

See how https://buildkite.com/automattic/wordpress-ios/builds/25994
failed:

```
error:
Build input file cannot be found:
'/opt/ci/builds/builder/automattic/wordpress-ios/DerivedData/Build/Intermediates.noindex/ArchiveIntermediates/WordPress Alpha/BuildProductsPath/Release-Alpha-iphoneos/WordPressDataObjC.o'.
Did you forget to declare this file as an output of a script phase or custom build rule which produces it? (in target 'WordPress' from project 'WordPress')
```

mokagio committed:
Replace all `CoreDataStack.h` imports with `WordPressDataObjc`

Not sure how it was working before — and in CI too, see
https://buildkite.com/automattic/wordpress-ios/builds/25995

My only guess is that the file became available in the DerivedData
somehow because the Swift package importing it was built before the app
targets?

Odd...

mokagio committed:
Move `CoreDataService` to `WordPressData`

mokagio committed:
Remove `Dummy` from `WordPressDataObjC` — No longer needed

It was needed just because there was no file in the package yet, but
we've addressed it whenw e moved `CoreDataService` in there.

mokagio commented on WordPress/Classes/Services/MediaImageService.swift:
```diff
@@ -1,7 +1,8 @@
-import UIKit
+import AsyncImageKit
 import CoreData
+import UIKit
+import WordPressData
 import WordPressShared
-import AsyncImageKit
```
In hindsight, sorting the imports alphabetically in the context of this PR makes the diff harder to read. Apologies.

kean commented:
> Ditch the logs, add it as a dependency, or implement an abstraction and injection system like we did for WordPressAuthenticator?

I would add `CocoaLumberjack` as a dependency. `WordPressAuthenticator` has an abstraction because it used to be in a separate repo and used by another app (s). If we want an abstraction, I suggest https://github.com/apple/swift-log but not in this project's scope.

> Some of the code uses WPException. Make it into its own package to import it in WordPressData and in the other code using WPException, or expose it through WordPressData and make those types that need it import WordPressData.

I suggest moving it to `WordPressSharedObjC`. This a small package with helper types like this. I'd say it's a no-goal to introduce too many new packages as it can make it harder to maintain and could impact compile time in a negative way. `@_exported` in an umbrella framework is, of course, also always an option.

> As far as I know we don't use those logs much.

Your call. If it's not logging anything you consider useful, I'd also remove it.

kean commented:
Btw, I really like the name "WordPressData" – rhymes with "CoreData".

mokagio commented:
> Btw, I really like the name "WordPressData" – rhymes with "CoreData".

Glad you like it 😄

mokagio commented:
> I suggest moving it [`WPException`] to WordPressSharedObjC

Cool. Done via https://github.com/wordpress-mobile/WordPress-iOS/pull/24157

kean reviewed (approved):
LGTM. Smoke tested.

mokagio committed:
Merge remote-tracking branch 'origin/trunk' into mokagio/wordpress-data-skeleton
